[PXCT-1519] - Nano R4 New Tutorial (External Interrupts) (#2677)
arduino/docs-content: content/hardware/03.nano/boards/nano-r4/tutorials/04.external-interrupts/content.md

***__Important note__: Avoid using interrupts for operations that are inherently slow, such as analog readings (ADC conversion takes time), complex mathematical calculations, or serial printing. These belong in your main loop where timing is less critical.***

### Polling vs Interrupts

To understand the value of interrupts, let's compare them with the traditional polling approach using a practical example: detecting a button press.

#### Polling Approach

With polling, your program must continuously check the button state in each loop cycle. This consumes processor cycles even when nothing is happening and can miss fast events if the program is busy with other tasks.

```arduino
// Polling approach - constantly checking
void loop() {
  // Check button state every loop iteration
  if (digitalRead(buttonPin) == LOW) {
    handleButtonPress();
  }
  
  // Other tasks that take time
  readSensors();        // Takes 20ms
  updateDisplay();      // Takes 30ms
  processData();        // Takes 15ms
  // Total loop time: 65ms+ means button must be held for at least that long
}
```

> The problem becomes clear when you consider timing. If your loop takes 65 milliseconds to complete, a quick button tap of 30 milliseconds will be missed entirely. Even worse, the processor wastes energy checking the button thousands of times when it is not being pressed.

Reviewer: ```suggestion
The problem becomes clear when you consider timing. If your loop takes 65 milliseconds to complete, a quick button tap of 30 milliseconds would be missed entirely. Even worse, the processor wastes energy checking the button thousands of times when it is not being pressed.

```
Minor suggestion

- **Never call blocking functions**: Functions like `delay()`, `Serial.print()`, `millis()`, or anything that waits for hardware are forbidden in ISRs. These functions often rely on interrupts themselves and calling them from an ISR can cause deadlocks or crashes.
- **Protect multi-byte shared data**. When your main code needs to read a multi-byte variable that an ISR modifies (like `long` or `float`), you must temporarily disable interrupts to prevent the value from changing mid-read. 

Here is how to safely read a multi-byte variable that is modified by an ISR:

```arduino
// Reading a 4-byte counter safely
volatile unsigned long pulseCount = 0;

void countISR() {
  pulseCount++;  // ISR modifies this
}

void loop() {
  // Wrong way - could read corrupted value
  unsigned long count = pulseCount;  // Dangerous!
  
  // Right way - atomic read
  noInterrupts();
  unsigned long count = pulseCount;  // Safe
  interrupts();
  
  // Now use 'count' for calculations
}
```
### Common Mistakes to Avoid

Even experienced developers fall into these traps when working with interrupts. Understanding these common mistakes will save you hours of debugging:

> - **Channel conflicts on the Nano R4 board**: The most frustrating error is when interrupts mysteriously stop working because you have accidentally used two pins on the same channel.The second attachment silently overwrites the first with no warning:

Reviewer: ```suggestion
- **Channel conflicts on the Nano R4 board**: The most frustrating error is when interrupts mysteriously stop working because you have accidentally used two pins on the same channel. The second attachment silently overwrites the first with no warning:

```
Minor suggestion

|     Program crashes or freezes     |        Blocking function called in ISR       |                                         Remove all `delay()`, `Serial.print()`, and similar functions from ISR                                         |
|      Incorrect counter values      |    Race condition on multi-byte variables    |                                           Use `noInterrupts()`/`interrupts()` when reading multi-byte values                                           |

## Conclusions

External interrupts on the Nano R4 board provide powerful event-driven capabilities for your projects. By understanding the channel limitations, following best practices for ISR implementation, and avoiding common pitfalls, you can build responsive applications that react instantly to real-world events. Remember: keep ISRs minimal, use `volatile` for shared variables, and carefully manage the 9 shared interrupt channels among the 13 available pins in the Nano R4 board.

## Next Steps

> Now that you understand interrupts on the Nano R4 board, consider building practical applications like tachometers, frequency counters, or power-efficient sensors that wake from sleep mode. Interrupts work with other Nano R4 features, such as the CAN bus for automotive projects or the DAC for triggered analog outputs.Check out the [Nano R4 User Manual](https://docs.arduino.cc/tutorials/nano-r4/user-manual/) to learn about its advanced features including the DAC, CAN bus, operational amplifiers, and HID capabilities.

Reviewer: ```suggestion
Now that you understand interrupts on the Nano R4 board, consider building practical applications, such as tachometers, frequency counters, or power-efficient sensors that wake from sleep mode. Interrupts integrate seamlessly with other Nano R4 features, such as the CAN bus for automotive projects or the DAC for triggered analog outputs. Check out the [Nano R4 User Manual](https://docs.arduino.cc/tutorials/nano-r4/user-manual/) to learn about its advanced features, including the DAC, CAN bus, operational amplifiers, and HID capabilities.

```
Minor suggestion

> For more advanced projects and examples specifically designed for the Nano R4, visit the [Arduino Project Hub](https://projecthub.arduino.cc/). You might also want to explore the [Arduino Forum's Nano R4 section](https://forum.arduino.cc/c/official-hardware/nano-family/nano-r4/221) to connect with other users and share your experiences external interrupts.

Reviewer: ```suggestion
For more advanced projects and examples specifically designed for the Nano R4, visit the [Arduino Project Hub](https://projecthub.arduino.cc/). You might also want to explore the [Arduino Forum's Nano R4 section](https://forum.arduino.cc/c/official-hardware/nano-family/nano-r4/221) to connect with other users and share your experiences with external interrupts.
```
Minor suggestion
